cmd/manager/manager.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	new, ok := n.(func(*logging.Logger) transport.Transport)
+	if !ok {
+		return fmt.Errorf("plugin %s constructor 'New' did not return type 'transport.Transport'", name)
+	}
+
+	transports[name] = new(logger)
+
+	c, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed parsing transport config for '%s'", name)
+	}
+
+	err = transports[name].Config(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//InitApplication initialize application plugin with configuration
+func InitApplication(name string, config interface{}) error {
+	n, err := initPlugin(name)
+	if err != nil {
+		return errors.Wrap(err, "failed initializing application plugin")
+	}
+
+	new, ok := n.(func(*logging.Logger) application.Application)
+	if !ok {
+		return fmt.Errorf("plugin %s constructor 'New' did not return type 'application.Application'", name)
+	}
+
+	applications[name] = new(logger)
+
+	c, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed parsing application plugin config for '%s'", name)
+	}
+
+	err = applications[name].Config(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//SetTransportHandlers load handlers binaries for transport
+func SetTransportHandlers(name string, handlerNames []string) error {
+	for _, hName := range handlerNames {
+		n, err := initPlugin(hName)
+		if err != nil {
+			return errors.Wrap(err, "failed initializing handler")
+		}
+
+		var hType string
+		switch new := n.(type) {
+		case (func() handler.MetricHandler):
+			metricHandlers[name] = append(metricHandlers[name], new())
+			hType = "MetricHandler"
+		case (func() handler.EventHandler):
+			eventHandlers[name] = append(eventHandlers[name], new())
+			hType = "EventHandler"
+		default:
+			return fmt.Errorf("handler %s constructor did not return type handler.EventHandler or handler.MetricsHandler", hName)
+		}
+		logger.Metadata(logging.Metadata{"handler": hName, "type": hType})
+		logger.Info("initialized handler")
+	}
+	return nil
+}
+
+//RunTransports spins off tranpsort + handler processes
+func RunTransports(ctx context.Context, wg *sync.WaitGroup, done chan bool) {
+	for name, t := range transports {
+		wg.Add(1)
+		go t.Run(ctx, wg, func(blob []byte) {
+			for _, handler := range metricHandlers[name] {
+				res := handler.Handle(blob)
+				if res != nil {
+					metricBus.Publish(res)
+				}
+			}
+			for _, handler := range eventHandlers[name] {
+				res, err := handler.Handle(blob)
+				if err != nil {
+					logger.Metadata(logging.Metadata{"error": err})
+					logger.Error("failed handling message")
+					continue
+				}
+				eventBus.Publish(res)
+			}
+		}, done)
+	}
+}
+
+//RunApplications spins off application processes
+func RunApplications(ctx context.Context, wg *sync.WaitGroup, done chan bool) {
+	for _, a := range applications {
+		eChan := make(chan data.Event)
+		mChan := make(chan []data.Metric)
+
+		eventBus.Subscribe(eChan)
+		metricBus.Subscribe(mChan)
```

Comment:
From my reading the cost is speed. AFAIR the more readers on a channel the slower it gets. I didn't do any benchmarking though, can find the article about it I guess. We might not see any performance issues currently, but if will will, subscribing only where is necessary is first thing which might help with that.

Reply:
I would suggested only subscribing to the bus(es) you need to use.  The other consideration would be to use buffered channels and send in bulk (i.e. group multiple metrics into a single chunk).

Reply:
Having multiple processes lread on one channel isn't an issue here. There is only ever one application process listening on any one channel at a time, since the [bus object](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/pkg/bus/bus.go#L39) accepts a new channel object every time Subscribe() is called. The bus object does a multicast on all of the subscribed channels whenever Publish() is called. Run() is still implemented poorly. Switching between channels can cause slowdowns if a lot of messages are coming from both and it still enables the problem of accidentally locking up one of the buses if the plugin forgets to listen on both. A buffered channel would eventually fill up and block in this circumstance as well. This was one of those things that I said "lll deal with later" and forgot about :). 

What could be done is give the Application plugin both the metric and event bus and let it decide which to subscribe to. The only reason I did not do this earlier was because I was not sure that the Applications should have to understand the internal bus structure, but it really doesn't matter. Or, instead of giving the plugins channels to listen on, the manager could expect the implementation of a separate callback for metrics and events. If I did this, it may be worth switching the bus structure over to callbacks as well.

Reply:
Please see my comment in bus.go about non-blocking sends... @pleimer